Description: avoid crashes caused by elaboration order.
 After binding the program with -p (pessimize the elaboration order)
 Ludovic discovered some elaboration order issues that are the cause
 for some of GPS's crashes at run-time.  This patch fixes these
 issues.
Last-Update: 2014-05-24
Author: Ludovic Brenta <lbrenta@debian.org>
Author: Nicolas Boulenguez <nicolas@debian.org>

--- a/common/core/src/string_hash.ads
+++ b/common/core/src/string_hash.ads
@@ -30,6 +30,7 @@
    --  Whether keys are case-sensitive
 
 package String_Hash is
+   pragma Elaborate_Body;
 
    type Name_Htable_Num is new Natural range 0 .. 6150;
    --  ??? This limitation should be raised (could be through use of
--- a/common/core/src/generic_stack.ads
+++ b/common/core/src/generic_stack.ads
@@ -18,6 +18,8 @@
 generic
    type Generic_Type is private;
 package Generic_Stack is
+   pragma Elaborate_Body;
+
    type Simple_Stack is private;
 
    type Generic_Type_Access is access all Generic_Type;
--- a/common/core/src/generic_list.ads
+++ b/common/core/src/generic_list.ads
@@ -26,6 +26,7 @@
    --  Free any dynamic memory associated with Data
 
 package Generic_List is
+   pragma Elaborate_Body;
 
    type List is private;
    type List_Node is private;
--- a/common/core/src/list_utils.ads
+++ b/common/core/src/list_utils.ads
@@ -18,6 +18,7 @@
 with Generic_List;
 
 package List_Utils is
+   pragma Elaborate_Body;
 
    generic
       with package List_Pkg is new Generic_List (<>);
--- a/common/core/src/file_utils.ads
+++ b/common/core/src/file_utils.ads
@@ -20,6 +20,7 @@
 with GNATCOLL.VFS; use GNATCOLL.VFS;
 
 package File_Utils is
+   pragma Elaborate_Body;
 
    function Suffix_Matches
      (File_Name : Filesystem_String;
--- a/kernel/src/gps-kernel.ads
+++ b/kernel/src/gps-kernel.ads
@@ -58,6 +58,8 @@
 with GPS.Core_Kernels;
 with GPS.Messages_Windows; use GPS.Messages_Windows;
 
+pragma Elaborate (GNATCOLL.VFS);
+
 package GPS.Kernel is
 
    type Kernel_Handle_Record is
--- a/common/ui/src/gui_utils.ads
+++ b/common/ui/src/gui_utils.ads
@@ -61,6 +61,7 @@
 with String_List_Utils;
 
 package GUI_Utils is
+   pragma Elaborate_Body;
 
    function Query_User
      (Parent        : Gtk.Window.Gtk_Window;
--- a/gvd/gvd/gvd-generic_view.ads
+++ b/gvd/gvd/gvd-generic_view.ads
@@ -70,6 +70,8 @@
    --  Return the kernel
 
 package GVD.Generic_View is
+   pragma Elaborate_Body;
+
    type Process_View_Record is abstract new Base_Type with private;
 
    function Get_Process
--- a/kernel/src/generic_views.ads
+++ b/kernel/src/generic_views.ads
@@ -29,6 +29,7 @@
 with Gtkada.MDI;
 
 package Generic_Views is
+   pragma Elaborate_Body;
 
    -----------------
    -- View_Record --
--- a/codefix/core/src/codefix-formal_errors.ads
+++ b/codefix/core/src/codefix-formal_errors.ads
@@ -31,6 +31,8 @@
 with Projects;
 with GNATCOLL.VFS;
 
+pragma Elaborate (Codefix.Text_Manager);
+
 package Codefix.Formal_Errors is
 
    package Cursor_Lists is new Generic_List (File_Cursor);
--- a/completion/core/src/completion.ads
+++ b/completion/core/src/completion.ads
@@ -38,6 +38,8 @@
 with GNATCOLL.VFS;     use GNATCOLL.VFS;
 with Xref;
 
+pragma Elaborate (GNATCOLL.VFS);
+
 package Completion is
 
    type Completion_Proposal;
